### PR TITLE
Alter batchSize based on memory usage

### PR DESCRIPTION
**Pull request by lachlan1m**
TODO:

- [x] Implement being able to alter batchSize based on memory usage.
- [x] Make sure that the adjustments being made don't negatively impact performance more than necessary to prevent heap out of memory error.

### CODE REVIEW

**Comment by ianhe8x** [comment]
What's the usual heapUsed:heapTotal ratio? How did you decide the low/high threshold?

**Comment by ianhe8x** [comment]
Looks good, i suggest adding a minimum batchSize just in case the heap used isn't shrunken for some reasons.

File: packages/node/src/indexer/fetch.service.ts [hunk 1]
```diff
@@ -213,6 +215,50 @@ export class FetchService implements OnApplicationShutdown {
     await this.getBestBlockHead();
   }
 
+  @Interval(BLOCK_TIME_VARIANCE * 10000)
+  getMemoryUsage() {
+    const formatMemoryUsage = (data) =>
```
Comment (ianhe8x) [inline packages/node/src/indexer/fetch.service.ts:220]: these declaration can be escalated

File: packages/node/src/indexer/fetch.service.ts [hunk 2]
```diff
@@ -213,6 +215,50 @@ export class FetchService implements OnApplicationShutdown {
     await this.getBestBlockHead();
   }
 
+  @Interval(BLOCK_TIME_VARIANCE * 10000)
```
Comment (ianhe8x) [inline packages/node/src/indexer/fetch.service.ts:218]: The interval isn't related to block time logically. If you think 50s is a reasonable value, just make a const and assign 50,000 to it.

**Comment by stwiname** [comment]
Out of curiosity, what sort of batch size does it tend to and with how much memory?

**Comment by lachlan1m** [comment]
> What's the usual heapUsed:heapTotal ratio? How did you decide the low/high threshold?

It is usually quite high in the 0.8 - 0.95% range (heapUsed / heapTotal). This is how I came up with the threshold values.

**Comment by lachlan1m** [comment]
> Out of curiosity, what sort of batch size does it tend to and with how much memory?

It tended jump around quite a lot because memory usage can change sporadically. So what I've done is lengthen the interval in which will check memory and increase the amount scales down batch size by to 10% (if it gets above 95% heap usage).

**Comment by lachlan1m** [comment]
I've also added a flag which can be used to disable this feature in yargs (--scale-batch-size).

File: packages/node/src/yargs.ts [hunk 3]
```diff
@@ -39,6 +39,12 @@ export function getYargsOption() {
       describe: 'Batch size of blocks to fetch in one round',
       type: 'number',
     },
+    'scale-batch-size': {
+      type: 'boolean',
+      demandOption: false,
+      describe: 'scale batch size based on memory usage',
+      default: true,
```
Comment (ianhe8x) [inline packages/node/src/yargs.ts:46]: default to false

**Review by ianhe8x — APPROVED** [review]

**Pull request merged**